cmd/zc_filter.go

```diff
@@ -156,3 +174,38 @@ func buildIncludeFilters(patterns []string) []objectFilter {
 
 	return []objectFilter{&includeFilter{patterns: validPatterns}}
 }
+
+type filterSet []objectFilter
+
+// GetEnumerationPreFilter returns a prefix that is common to all the include filters, or "" if no such prefix can
+// be found. (The implementation may return "" even in cases where such a prefix does exist, but in at least the simplest
+// cases, it should return a non-empty prefix.)
+// The result can be used to optimize enumeration, since anything without this prefix will fail the filterSet
+func (fs filterSet) GetEnumerationPreFilter(recursive bool) string {
+	if recursive {
+		return ""
+		// we don't/can't support recursive cases yet, with a strict prefix-based search.
+		// Because if the filter is, say "a*", then, then a prefix of "a"
+		// will find: enumerationroot/afoo and enumerationroot/abar
+		// but it will not find: enumerationroot/virtualdir/afoo
+		// even though we want --include-pattern to find that.
+		// So, in recursive cases, we just don't use this prefix-based optimization.
+		// TODO: consider whether we need to support some way to separately invoke prefix-based optimization
+		//   and filtering.  E.g. by a directory-by-directory enumeration (with prefix only within directory),
+		//   using the prefix feature in ListBlobs.
```

Comment:
This might be difficult since directories don't actually exist in blob, and in order to discover them, we have to do a full flat listing anyway, iirc.

Reply:
There's a solution, with the Prefix field, I understand.